File: leetcode/path-sum-iii/path-sum-iii.rb
# LeetCode #437 - Path Sum III
# https://leetcode.com/problems/path-sum-iii/
# Prefix-sum hashing solution. Runs in linear time.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} target_sum
# @return {Integer}
def path_sum(root, target_sum)
  history = Hash.new(0)
  history[0] = 1

  dfs = ->(node, acc) do
    return 0 unless node

    acc += node.val
    count = history[acc - target_sum]

    history[acc] += 1
    count += dfs.call(node.left, acc)
    count += dfs.call(node.right, acc)
    history[acc] -= 1

    count
  end

  dfs.call(root, 0)
end

File: leetcode/path-sum-iii/path-sum-iii.c
// LeetCode #437 - Path Sum III
// https://leetcode.com/problems/path-sum-iii/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

static int dfs(const struct TreeNode *const root, int subtotal, const int total)
{
    if (!root) return 0;

    printf("%d %d\n", subtotal, total);

    subtotal += root->val;

    return (subtotal == total ? 1 : 0)
            + dfs(root->left, subtotal, total)
            + dfs(root->left, 0, total)
            + dfs(root->right, subtotal, total)
            + dfs(root->right, 0, total);
}

int pathSum(const struct TreeNode *const root, int targetSum)
{
    return dfs(root, 0, targetSum);
}
